pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+
+
+Generic and recursive protocols
+-------------------------------
+
+Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
+and ``Iterator`` are generic protocols. They are defined similar to normal
+non-protocol generic types::
+
+  T = TypeVar('T', covariant=True)
+
+  class Iterable(Protocol[T]):
+      @abstractmethod
+      def __iter__(self) -> Iterator[T]:
+          ...
+
+Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``.
+
+Recursive protocols are also supported. Forward references to the protocol
+class names can be given as strings as specified by PEP 484. Recursive
+protocols will be useful for representing self-referential data structures
+like trees in an abstract fashion::
+
+  class Traversable(Protocol):
+      leafs: Iterable['Traversable']
+
+
+Using Protocols
+===============
+
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
+protocol types. For variables and parameters with protocol types, subtyping
+relationships are subject to the following rules:
+
+* A protocol is never a subtype of a concrete type.
+* A concrete type or a protocol ``X`` is a subtype of another protocol ``P``
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
+* Edge case: for recursive protocols, structural subtyping is decided
+  positively for situations where such decision depends on itself. Continuing
+  the previous example::
+
+    class Tree(Generic[T]):
+        def __init__(self, value: T,
+                           leafs: 'List[Tree[T]]') -> None:
+            self.value = value
+            self.leafs = leafs
+
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
+
+Generic protocol types follow the same rules of variance as non-protocol
+types. Protocols can be used in all special type constructors provided
+by the ``typing`` module and follow the corresponding subtyping rules.
+For example, protocol ``PInt`` is a subtype of ``Union[PInt, int]``.
+
+
+``Union[]`` and ``All[]``
+-------------------------
+
+``Union`` of protocol classes behaves the same way as for non-protocol
+classes. For example::
+
+  from typing import Union, Optional, Protocol
+
+  class Exitable(Protocol):
+      @abstracrmethod
+      def exit(self) -> int:
+          ...
+  class Quitable(Protocol):
+      @abstractmethod
+      def quit(self) -> Optional[int]:
+          ...
+
+  def finish(task: Union[Exitable, Quitable]) -> int:
+      ...
+  class GoodJob:
+      ...
+      def quit(self) -> int:
+          return 0
+  finish(GoodJob()) # OK
+
+In addition, we propose to add another special type construct
+``All`` that represents intersection types. Although for normal types
```

Comment:
I think we could include this for three reasons:
* I could easily imagine a function parameter that requires more than one protocol, especially if the preferred style is to have several simple protocols, rather than few complex. By the way, some classes in ``collections.abc`` are simple intersections, for example ``Collection == All[Sized, Iterable, Container]``.
* It could be easily implemented for protocols. I also propose to use ``All`` instead of ``Intersection`` to emphasize it only works for protocols (mnemonics: variable implements _all_ these protocols). On the contrary, I think that intersections of nominal classes would be _very_ rarely used.
* If we decide to include this later, then people might simply be not aware of this feature (this is a minor reason, but still we need to take this into account).